solr/solrj/src/java/org/apache/solr/common/util/CommonTestInjection.java

```diff
@@ -75,4 +78,49 @@ public static boolean injectDelay() {
     }
     return true;
   }
+
+  /**
+   * This should ONLY be set from unit test cases.
+   *
+   * Block at certain code execution point (defined by the key at which injectWaitForCondition was inserted)
+   * until the provided condition is fulfilled
+   * 
+   * @see CommonTestInjection#injectWaitForCondition(String)
+   * @param key
+   * @param condition
+   */
+  public static void setWaitForCondition(String key, Condition condition) {
+    synchronized (waitForConditions) {
+      if (condition != null) {
+        waitForConditions.put(key, condition);
+      } else {
+        waitForConditions.remove(key);
+      }
+    }
+  }
+
+  /**
+   * This injects a precise wait until a condition is fulfilled, such condition is looked up by the corresponding key.
+   *
+   * The key could simply be the fully qualified class name or could be more granular like class name + other id
+   * (such as method name) and should only be set by corresponding unit test cases with setWaitForCondition
+   *
+   * Therefore, in non-testing environment, this should always return w/o blocking since the lookup map should be empty.
+   *
+   * @see CommonTestInjection#setWaitForCondition(String, Condition) 
+   * @param key
+   */
+  public static void injectWaitForCondition(String key) {
+    Condition condition = waitForConditions.get(key);
+    if (condition != null) {
+      condition.check();
+    }
+  }
+
+  public interface Condition {
```

Comment:
`Condition` here immediately makes me. think `java.util.concurrent.locks.Condition` -- do we need this at all? Could just be a runnable, which is actually more semantically transparent based on the way we're using it here?

And if so, maybe rename methods to `inject(String key)` and `setInject(String key, Runnable inject)`

Reply:
I probably have considered that (i recall perhaps looking for suitable functional interface), and might or might not have thought of `Runnable` (there goes my goldfish memory lol)

I have similar sentiments with some people here https://stackoverflow.com/questions/23958814/functional-interface-that-takes-nothing-and-returns-nothing#comment52094675_23958916 that runnable does not appear to describe the case very well ( I do have the tendency to think that it's going to spawn a new thread or something 😅 ). So if i might misread that as inject certain background things to executed. 

Another reason that deters me sometimes to use a very generic interface is the difficulty to make code references, cause scan would give me like thousands of results for those common interface 😹 So if the interface is super simple, i do have the tendency to create a custom one, if it's bit more descriptive.

That being said, `Condition` is NOT a good choice either 😁 as it confuses you right the way and i don't think it describes what we want that well either. Thoughts? 💡

Reply:
Renamed it to `Breakpoint`, i think that might be a bit more suitable?

Reply:
Good point about not using `Runnable` -- that's bitten me before (not being able to effectively search for examples/instances).

`Breakpoint` seems good to me -- this is probably good to try to upstream, at least to get feedback on the general approach. Probably doesn't make too much sense to put more effort into workshopping the class name before upstreaming, since we'll probably discuss it then as well?

Reply:
For sure, created ticket in apache jira https://issues.apache.org/jira/browse/SOLR-16696 !